piptools/writer.py

```diff
@@ -239,11 +239,20 @@ def _format_requirement(
 
         line = format_requirement(ireq, marker=marker, hashes=ireq_hashes)
 
-        if not self.annotate or key_from_ireq(ireq) in primary_packages:
+        if not self.annotate:
             return line
 
-        # Annotate what packages this package is required by
-        required_by = reverse_dependencies.get(ireq.name.lower(), [])
+        # Annotate what packages or reqs-ins this package is required by
+        required_by = reverse_dependencies.get(ireq.name.lower(), set())
```

Comment:
```python
required_by = set()
if hasattr(ireq, "_source_ireqs"):
    required_by |= {
        _comes_from_as_string(src_ireq) for src_ireq in ireq._source_ireqs
    }
elif ireq.comes_from:
    required_by.add(_comes_from_as_string(ireq))

...

def _comes_from_as_string(ireq):
    if isinstance(ireq.comes_from, six.string_types):
        return ireq.comes_from
    return key_from_ireq(ireq.comes_from)
```

I came up with this idea which also fixes #293. How do you like it?

Reply:
If it fixes #293, that's excellent!

1. Should `_comes_from_as_string` live in `writer`, or `utils`?
2. To what extent can/should `reverse_dependencies` and `primary_packages` be removed? From `OutputWriter` completely, but left alone in `cache` and `resolver`? A bunch of writer tests currently use those.

Reply:
1. Both options fine by me
2. Remove `reverse_dependencies` and `primary_packages` from `OutputWriter` completely. It would affect the `piptools.compile` module and related tests.

Reply:
This required many changes in the tests, and a `None` check filter in that `required_by` comprehension. Please have a look at the new state.

Questions:

- In `compile.py`, we stopped passing around `reverse_dependencies`, so I removed its 
  creation. Is that appropriate, or is some beneficial cache behavior lost?
- Do the comments there (`"Compute reverse dependency annotations statically . . ."`) need updating/deleting?

Reply:
> In `compile.py`, we stopped passing around `reverse_dependencies`, so I removed its
>   creation. Is that appropriate, or is some beneficial cache behavior lost?

It's appropriate.

> Do the comments there (`"Compute reverse dependency annotations statically . . ."`) need updating/deleting?

IMO this whole comment block could be deleted because it wouldn't be relevant anymore.

Reply:
:+1: Thanks